File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        min_price = prices[0]
        profit = 0
        for i in prices:
            # agar price minprice se choti hai to usko minprice krlo
            if i<min_price:  
                min_price = i
                # profit = sell price - minprice
            profit = max(profit , i - min_price)

        return profit
        

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // Initialize minPrice to the first element
        int minPrice = prices[0];
        
        // Initialize maxProfit to 0 (no profit if no valid transaction possible)
        int maxProfit = 0;
        
        // Traverse through the array once - O(n) time complexity
        for (int i = 1; i < prices.length; i++) {
            // Calculate current profit if we bought at minPrice and sell today
            int currentProfit = prices[i] - minPrice;
            
            // Update maxProfit if current profit is higher
            maxProfit = Math.max(maxProfit, currentProfit);
            
            // Update minPrice if current price is lower than what we saw
            minPrice = Math.min(minPrice, prices[i]);
        }
        
        return maxProfit;
    }
}